File: 56. Merge Intervals.cc
// TC: O(nlogn)
// SC: O(1)

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
 
        sort(intervals.begin(), intervals.end());
        int n = intervals.size();
        vector<vector<int>> ans;
        ans.reserve(5);
        ans.push_back(intervals[0]);
        for (int i = 1; i < n; i++) {
            if (ans.back()[1] >= intervals[i][0])
                ans.back()[1] = max(ans.back()[1], intervals[i][1]);
            else
                ans.push_back(intervals[i]);
        }
        
        return ans;        
    }
};

File: java/56. Merge Intervals.java
// TC: O(N log N) for sorting + O(N) for merging = O(N log N) overall, where N is the number of intervals
// SC: O(1) excluding the space for output
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        List<int[]> res = new ArrayList<>();
        
        for (int[] interval : intervals) {
            if (res.isEmpty() || res.get(res.size() - 1)[1] < interval[0]) {
                res.add(interval);  // 不重疊，添加新區間
            } else {
                res.get(res.size() - 1)[1] = Math.max(res.get(res.size() - 1)[1], interval[1]);  // 重疊，合併
            }
        }
        
        return res.toArray(new int[res.size()][]);
    }

}